### ISSUE TITLE
Flowable Timer Start Event not auto deleted when there is new deployment version 

### ISSUE THREAD

**Issue by liang-faan** [flowable/flowable-engine#3403]
```text
*Timer Start Event*
Description
A timer start event is used to create process instances at given time. It can be used both for processes that should start only once and for processes that should start in specific time intervals.
Note: a sub-process cannot have a timer start event.
Note: a start timer event is scheduled as soon as process is deployed. There is no need to call startProcessInstanceByXXX, although calling start process methods is not restricted and will cause one more starting of the process at the time of startProcessInstanceByXXX invocation.
Note: when a new version of a process with a start timer event is deployed, the job corresponding with the previous timer will be removed. The reasoning is that normally it is not desirable to keep automatically starting new process instances of the old version of the process.
```
Hi Flowable,
I impemented a start timer event with recurring schedule, but I found that the older version timer job was not auto removed. it causes there are some timer jobs in the flowable engine running concurrently. but according to above "Note", the start timer job will be replaced by newer version. 

    <startEvent id="PROWATCH_INPUT_MONITOR_START" name="PROWATCH_INPUT_MONITOR_START" isInterrupting="false">
      <timerEventDefinition>
        <timeCycle>0 0/5 * * * ?</timeCycle>
      </timerEventDefinition>
    </startEvent>

**Comment by liang-faan** [comment]
I am using v6.7.2

**Comment by liang-faan** [comment]
`
/**
     * Updates all timers and events for the process definition. This removes obsolete message and signal subscriptions and timers, and adds new ones.
     */
    public void updateTimersAndEvents(ProcessDefinitionEntity processDefinition,
            ProcessDefinitionEntity previousProcessDefinition, ParsedDeployment parsedDeployment) {

        Process process = parsedDeployment.getProcessModelForProcessDefinition(processDefinition);
        BpmnModel bpmnModel = parsedDeployment.getBpmnModelForProcessDefinition(processDefinition);

        eventSubscriptionManager.removeObsoleteMessageEventSubscriptions(previousProcessDefinition);
        eventSubscriptionManager.removeObsoleteSignalEventSubScription(previousProcessDefinition);
        eventSubscriptionManager.removeObsoleteEventRegistryEventSubScription(previousProcessDefinition);
        eventSubscriptionManager.addEventSubscriptions(processDefinition, process, bpmnModel);

        timerManager.removeObsoleteTimers(processDefinition);
        timerManager.scheduleTimers(processDefinition, process);
    }

`

this  **timerManager.removeObsoleteTimers(processDefinition);** shouldn't  be  _timerManager.removeObsoleteTimers(**previousProcessDefinition**);_

**Comment by jbarrez** [comment]
It would be more logical, yes. However, it doesn't change anything, as only the tenantId and key are taken from the passed process definition (the findJobsByTypeAndProcessDefinitionKeyAndTenantId method is used). 

This is also what is tested in the StartTimerEventTest linked above.

**Comment by liang-faan** [comment]
Hi Jbarrez,

thanks for the quick response. 
but I don't think it works. 
in my environment, I am using findJobsByTypeAndProcessDefinitionKeyNoTenantId which query as blow. 

`
<select id="selectTimerJobByTypeAndProcessDefinitionKeyNoTenantId" parameterType="org.flowable.common.engine.impl.db.ListQueryParameterObject" resultMap="jobResultMap">
		select J.*
		from ${prefix}ACT_RU_TIMER_JOB J
		**inner join ${prefix}ACT_RE_PROCDEF P on J.PROC_DEF_ID_ = P.ID_**
		where J.HANDLER_TYPE_ = #{parameter.handlerType}
		and P.KEY_ = #{parameter.processDefinitionKey}
		and (P.TENANT_ID_ = '' or P.TENANT_ID_ is null)
	</select>
`

above PROC_DEF_ID_ in my environment actually follow by **WorkFlowKey:<UUID>** which it is unique ID for each version definition. if using current deployment process definition, it wont able to fine previously version job.

**Comment by jbarrez** [comment]
The PROC_DEF_ID_ is used in the inner join, it is not passed into the query. So this should work, regardless of what you've set in the ID.

In Flowable, each process definition version shares the same unique key. E.g. v1, v2, v3, ... all share the same process definition key. This is fundamental to how Flowable works. Are you saying your two process definitions have different keys?

Can you provide a unit test showing the problem?

**Comment by liang-faan** [comment]
Hi Jbarrez, 

Sorry, it is not the issue of selectTimerJobByTypeAndProcessDefinitionKeyNoTenantId.
You are right. After debugging, the query it works. 
But I found the issue actually inside CancelJobsCmd - 

```java
@Override
    public Void execute(CommandContext commandContext) {
        JobEntity jobToDelete = null;
        for (String jobId : jobIds) {
            jobToDelete = commandContext
                    .getJobEntityManager()
                    .findJobById(jobId);

            if (jobToDelete != null) {
                if (commandContext.getProcessEngineConfiguration().getEventDispatcher().isEnabled()) {
                    commandContext.getProcessEngineConfiguration().getEventDispatcher().dispatchEvent(
                            ActivitiEventBuilder.createEntityEvent(FlowableEngineEventType.JOB_CANCELED, jobToDelete),
                            EngineConfigurationConstants.KEY_PROCESS_ENGINE_CONFIG);
                }

                jobToDelete.delete();

            } else {
                TimerJobEntity timerJobToDelete = commandContext.getTimerJobEntityManager().findJobById(jobId);

                if (timerJobToDelete != null) {
                    if (commandContext.getProcessEngineConfiguration().getEventDispatcher().isEnabled()) {
                        commandContext.getProcessEngineConfiguration().getEventDispatcher().dispatchEvent(
                                ActivitiEventBuilder.createEntityEvent(FlowableEngineEventType.JOB_CANCELED, timerJobToDelete),
                                EngineConfigurationConstants.KEY_PROCESS_ENGINE_CONFIG);
                    }

                    timerJobToDelete.delete();
                }
            }
        }
        return null;
    }

```

 TimerJobEntity timerJobToDelete = commandContext.getTimerJobEntityManager().findJobById(jobId);

above findJobById, I seeing in the query is calling "Select * from ACT_RU_JOB", but because of my start event is scheduled based on timecycle, And there is no running job triggered during deployment. 
which means above findJobById actually returns nothing. 

Finally became no job be canceled in CancelJobsCmd

**Comment by filiphr** [comment]
Hey @liang-faan,

I think that Joram shared some of our tests that are testing exactly the functionality you are saying that isn't working. The queries in the examples you shared are either checking the runtime table or the timer table. The job will be deleted in either case.

Can you please provide a minimal reproducable example showing the problem you are facing. Without it, we cannot reproduce the problem you are having.

**Comment by liang-faan** [comment]
Hi @filiphr,

After further debugging and trying out different scenarios, we found the issue is happened on overdue timer jobs. 

The scenario is 
- There is some overdue or nearer over due jobs in table ACT_RU_TIMER_JOB
- then server restart to deploy newer version flows
- we found that those overdue jobs were picked up by flowable-bpm-acquire-timer-jobs before server proceed newer version flows 
- Then when come to new flows deployment, as perviously overdue jobs were deleted from table ACT_RU_TIMER_JOB, there was no jobs be canceled by flowable deployment, but deployment thread inserted the newer version timer job
- however, after those older jobs complete process, they proceed insert back to table ACT_RU_TIMER_JOB

because of above process, it caused multiple version timer jobs remains in ACT_RU_TIMER_JOB. 

is there anyway postpone those flowable async jobs startup after newer flow deployment complete?
